### PR TITLE
mgr/dashboard: fix bucket versioning when locking is enabled

### PR DESCRIPTION
**Pull request by avanthakkar**
This PR fixes issue of suspending bucket versioning even if object locking is enabled. As per this 
https://github.com/ceph/ceph/blob/f7294bca9d392ff58a8854902e4c370dc693205b/src/rgw/rgw_op.cc#L2538 , it seems like if object locking is enabled then no other status of versioning should be allowed other than `Enabled`.

Fixes: https://tracker.ceph.com/issues/50545
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

Before:
Screenshot from 2021-04-28 00-49-37
After:
Screenshot from 2021-05-03 15-06-44

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test api

**Review by alfonsomthd — CHANGES_REQUESTED** [review]
You're addressing the issue only at the UI level. The Single Page App is only one of the potential consumers of the API, so this issue has to be addressed also at the API level.

**Comment by avanthakkar** [comment]
> You're addressing the issue only at the UI level. The Single Page App is only one of the potential consumers of the API, so this issue has to be addressed also at the API level.

I think backend already takes care of it, bcz if you try to `PUT` bucket with `versioning!=Enabled` along with object locking enabled then it throws error with status code 409, For e.g.:
```
PUT | https://localhost:11000/api/rgw/bucket/abc?daemon_name=8000&bucket_id=05257eb4-8a5b-4fbc-b9d9-ba74deb1ecaa.4324.1&uid=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef&versioning_state=Suspended&mfa_delete=Disabled&mfa_token_serial=&mfa_token_pin=&lock_mode=COMPLIANCE&lock_retention_period_days=3&lock_retention_period_years=0
```
Response:
```
detail: RGW REST API failed request with status code 409
(b'{"Code":"InvalidBucketState","Message":"bucket versioning cannot be disabled'
 b' on buckets with object lock enabled","BucketName":"abc","RequestId":"tx0000'
 b'0000000000000001e-0060891400-10de-default","HostId":"10de-default-default"}')
 ```

**Comment by avanthakkar** [comment]
> 0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef

@alfonsomthd I got the idea now, thanks! This is what it looks like if we try to send a invalid `PUT`. I've shown same status code `409` as rgw component returns.
Screenshot from 2021-04-28 15-39-16

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 1]
```diff
@@ -300,12 +300,16 @@ def set(self, bucket, bucket_id, uid, versioning_state=None,
         uid_tenant = uid[:uid.find('$')] if uid.find('$') >= 0 else None
         bucket_name = RgwBucket.get_s3_bucket_name(bucket, uid_tenant)
 
+        # Update locking if it is enabled.
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/controllers/rgw.py:303]: Provided the code that this comment is talking about is 10 lines below why not just putting the 
```
if locking['lock_enabled']:
            self._set_locking(uid, daemon_name, bucket_name, lock_mode,
                              lock_retention_period_days,
```
logic just before ``versioning_state``` code so the comment and the code are together?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 2]
```diff
@@ -116,7 +116,7 @@
         </div>
 
         <!-- Versioning -->
-        <fieldset *ngIf="editing">
+        <fieldset *ngIf="editing && !bucketForm.getValue('lock_enabled')">
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:119]: Just a question: would it be more clear for the user if we show the versioning state but in read-only mode (not able to uncheck) so he clearly understands that object locking implies enabling versioning ?

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html]: Yes makes sense. I've addressed your suggestion, thanks!

File: src/pybind/mgr/dashboard/controllers/rgw.py (hunk 1 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/rgw.py]: I instead moved the comment below.

**Comment by avanthakkar** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 3]
```diff
@@ -120,7 +120,28 @@
           <legend class="cd-header"
                   i18n>Versioning</legend>
 
-          <div class="form-group row">
+          <div *ngIf="bucketForm.getValue('lock_enabled'); then disableVersioning else enableVersioning"
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:123]: You don't need to create 2 templates to disable (or not) versioning:
```
--- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts
@@ -147,6 +147,9 @@ export class RgwBucketFormComponent extends CdForm implements OnInit {
             this.isVersioningAlreadyEnabled = this.isVersioningEnabled;
             this.isMfaDeleteAlreadyEnabled = this.isMfaDeleteEnabled;
             this.setMfaDeleteValidators();
+            if (value['lock_enabled']) {
+              this.bucketForm.controls['versioning'].disable();
+            }
           }
         }
```

File: src/pybind/mgr/dashboard/frontend/cypress/integration/rgw/buckets.po.ts [hunk 4]
```diff
@@ -32,52 +36,83 @@ export class BucketsPageHelper extends PageHelper {
...
-      .as('bucketDataTable');
+      // Check versioning enabled:
+      cy.get('@bucketDataTable').find('tr').its(2).find('td').last().should('have.text', new_owner);
+      cy.get('@bucketDataTable').find('tr').its(11).find('td').last().as('versioningValueCell');
 
-    // Check versioning enabled:
-    cy.get('@bucketDataTable').find('tr').its(2).find('td').last().should('have.text', new_owner);
-    cy.get('@bucketDataTable').find('tr').its(11).find('td').last().as('versioningValueCell');
+      return cy.get('@versioningValueCell').should('have.text', this.versioningStateEnabled);
+    } else {
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/cypress/integration/rgw/buckets.po.ts:81]: As the "if" clause is ending wit ha return, it seems you don't need the "else" clause.

**Review by alfonsomthd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/cypress/integration/rgw/buckets.po.ts (hunk 4 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/cypress/integration/rgw/buckets.po.ts:81]: Yes, fixed it.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html (hunk 3 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html]: Thanks @alfonsomthd , applied the changes.

**Review by aaSharma14 — APPROVED** [review]
LGTM.!

**Review by alfonsomthd — APPROVED** [review]
LGTM Good work!

**Pull request merged**